### PR TITLE
Refactor & fix generator code

### PR DESCRIPTION
**Pull request by Arkay92**
# Detailed Improvements and Changes
Here's a detailed comparison of the original code and the refactored version with a focus on improvements and changes:

### Original
```
self = cls(fndesc.native, fndesc.modname, qualname, unique_name, fndesc.doc, fndesc.typemap, restype, fndesc.calltypes, args, fndesc.kws, argtypes=argtypes, mangler=mangler, inline=False, env_name=fndesc.env_name)
return self
```
### Refactored
```
return cls(fndesc.native, fndesc.modname, qualname, unique_name, fndesc.doc, fndesc.typemap, restype, fndesc.calltypes, args, fndesc.kws, argtypes=argtypes, mangler=mangler, inline=False, env_name=fndesc.env_name)
```
### Improvement 
Simplified the return statement by directly returning the newly created instance.

### Changed BaseGeneratorLower class from an old-style class to a new-style class:
```
class BaseGeneratorLower(object):  # Original
class BaseGeneratorLower:  # Refactored
```

### Improvement:
Consistency with modern Python class definitions.

### Consolidated Pointer Retrieval Methods:
Original:
```
def get_args_ptr(self, builder, genptr):
    return cgutils.gep_inbounds(builder, genptr, 0, 1)

def get_resume_index_ptr(self, builder, genptr):
    return cgutils.gep_inbounds(builder, genptr, 0, 0, name='gen.resume_index')

def get_state_ptr(self, builder, genptr):
    return cgutils.gep_inbounds(builder, genptr, 0, 2, name='gen.state')
```
Refactored:
```
def get_pointer(self, builder, genptr, *indices):
    return cgutils.gep_inbounds(builder, genptr, 0, *indices)
```
### Improvement
Combined similar methods into a single, more flexible method to reduce redundancy and improve maintainability.

### Refactored the lower_init_func method:

Original:
```
lower.debug_print("# low_init_func incref")
if self.context.enable_nrt:
    for argty, argval in zip(self.fndesc.argtypes, lower.fnargs):
        self.context.nrt.incref(builder, argty, argval)
```

Refactored:
```
self.incref_arguments(lower, builder)
```
### Improvement
Extracted incref logic into its own method (incref_arguments) for clarity and reusability.

### Extracted Reusable Methods:
Original:
```
# Part of lower_next_func
stop_block = function.append_basic_block("stop_iteration")
builder.position_at_end(stop_block)
self.call_conv.return_stop_iteration(builder)
```
Refactored:
```
stop_block = self.add_stop_iteration_block(builder, function)
```
### Improvement
Modularized the creation of the stop iteration block into its own method (add_stop_iteration_block), enhancing readability.

### Changed string formatting to f-strings:
Original:
```
block_name = "generator_resume%d" % (index,)
```
Refactored:
```
block_name = f"generator_resume{index}"
```
### Improvement
Used f-strings for improved readability and consistency with modern Python code.

## Consistent Debug Printing:

Original:
```
self.context.debug_print(builder, "DEBUGJIT: {0}".format(msg))
```
Refactored:
```
self.context.debug_print(builder, f"DEBUGJIT: {msg}")
Improvement: Consistent use of f-strings for debug printing.
```
### Removed Redundant Code
Original:
```
retty = self.context.get_return_type(self.gentype)
argsty = retty.elements[1]
statesty = retty.elements[2]
```
Refactored: 
Simplified and clarified by removing unused variables argsty and statesty in the context of their usage.

## Refactored Method Calls:

Original:
```
self.arg_packer.load_into(builder, self.get_args_ptr(builder, genptr), lower.fnargs)
```
Refactored:
```
self.arg_packer.load_into(builder, self.get_pointer(builder, genptr, 1), lower.fnargs)
```
### Improvement
Ensured consistent usage of the new get_pointer method.

## General Code Cleanup:
- Removed unnecessary comments and consolidated methods to improve overall readability and maintainability.
- Ensured that all methods have proper docstrings and inline comments where necessary.

## Summary of Improvements
- Refactoring and Modularization: Improved code maintainability and readability by refactoring redundant code into reusable methods.
- Modern Python Practices: Adopted modern Python practices like using f-strings and new-style class definitions.
- Consistency: Ensured consistent usage of methods and practices throughout the code.
- Enhanced Readability: Added detailed docstrings and inline comments for better understanding and future maintenance.

### CODE REVIEW

**Comment by esc** [comment]
@Arkay92 welcome to Numba and thank you for looking into this! 🙌

 I had a quick glance at it and we do prefer to have code `flake8` compliant. Maybe you could add the following patch and then reformat your changes, so as to be compliant.

```
diff --git i/.flake8 w/.flake8
index 5c2885a68e..0b123713e8 100644
--- i/.flake8
+++ w/.flake8
@@ -43,7 +43,6 @@ exclude =
     numba/misc/numba_entry.py
     numba/stencils/stencilparfor.py
     numba/core/ir.py
-    numba/core/generators.py
     numba/misc/appdirs.py
     numba/core/caching.py
     numba/core/debuginfo.py
```

Also, I see you are removing inline comments starting with a `#` -- please can you revert those where appropriate as they do help readers to understand some of the code involved. 

Lastly, I'm not sure if this is documented anywhere (and probably should be if it isn't) -- but we try to use the NumPyDoc standard for docstrings. It is is documented here:

https://numpydoc.readthedocs.io/en/latest/format.html

It would be good if the refactoring could adhere to this standard, thank you.

**Comment by gmarkall** [comment]
> I had a quick glance at it and we do prefer to have code flake8 compliant. Maybe you could add the following patch and then reformat your changes, so as to be compliant.

Maybe this should not be done until all review feedback is addressed and the PR would otherwise be good to merge - otherwise the PR might become very hard to review with a hugely noisy diff.

**Comment by esc** [comment]
@Arkay92 question: was a large language model LLM used during the preparation of this PR?

**Comment by esc** [comment]
@Arkay92 it seems like the refactor broke the tests, can you fix?

**Comment by esc** [comment]
> > I had a quick glance at it and we do prefer to have code flake8 compliant. Maybe you could add the following patch and then reformat your changes, so as to be compliant.
> 
> Maybe this should not be done until all review feedback is addressed and the PR would otherwise be good to merge - otherwise the PR might become very hard to review with a hugely noisy diff.

@gmarkall : with the amount of code churn, I think it is inevitable that this diff will be noisy.  What do you think, should we suggest to break the PR into smaller PRs?

File: numba/core/generators.py [hunk 1]
```diff
@@ -60,130 +89,204 @@ def __init__(self, lower):
...
+            The pointer to the generator structure.
+        indices : tuple
+            The indices to access the desired field in the structure.
+
+        Returns
+        -------
+        Value
+            The pointer to the specified field.
+        """
+        return cgutils.gep_inbounds(builder, genptr, 0, *indices)
```
Comment (esc) [inline numba/core/generators.py:129]: This is missing the `name` parameter entirely, I am not sure if that is a good idea.

**Comment by esc** [comment]
@Arkay92 thank you again for submitting this. After conferring with the other devs via chat, we have decided that we must reject this PR in it's current form. We are sorry about as you have already invested time into this. The reason is that this PR has gotten too big and out of hand for us to review it adequately, with 310 lines and several things being changed and all the tests are already breaking. We hope you understand.

We suggest that you break this PR into much smaller PRs, perhaps one for each refactoring category and method extraction and so on.

a) initial flake8 fixes and removing file from flake8 excludes
b) move to new style classes
c) use f-strings
d) fixup return statement in class
e) etc....

Does that make sense?

**Pull request closed without merging**